Handle `Span`s for byte and raw strings and add more detail

estebank opened the pull request:
CC #81208.

estebank committed:
Handle `Span`s for byte and raw strings and add more detail

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

estebank commented:
cc @wesleywiser

petrochenkov commented:
r? @petrochenkov

cjgillot commented on compiler/rustc_lexer/src/unescape.rs:
```diff
@@ -118,15 +118,15 @@ pub enum Mode {
-    RawStr,
-    RawByteStr,
+    RawStr(u16),
+    RawByteStr(u16),
```
What do these `u16` represent?

wesleywiser replied:
Looks like the number of `#` characters used in a raw string like (`r##"hello world"##`). A doc comment would be nice.

petrochenkov replied:
These numbers don't logically belong to `Mode` though.

Not sure why some of this span arithmetic is done outside of `emit_unescape_error` and then undone inside it, maybe it can be simplified, need to look in more detail.

estebank replied:
We need both the span for the outer span (`"..."`) and inner span (`...`). An alternative would be to pass those two spans into the function (they are both needed) to avoid carrying that information in the `Mode`.

estebank replied:
Changed to pass in both `Span`s (as both are needed for different errors).

petrochenkov commented:
I knew something was wrong here, the ICE doesn't happen because `#`s are counted incorrectly, the ICE happens because `emit_unescape_error` performs arithmetic on the overridden span (`StringReader::override_span`) which may be entirely arbitrary, but *happens* to point to the full byte string literal in this case.
So this change fixing the issue is a pure accident, and for other `override_span`s (e.g. from proc macros) it will still ICE.

petrochenkov commented:
The linked issue is fixed in https://github.com/rust-lang/rust/pull/81337 by avoiding reparsing values in key-value attributes for the second time (that's where the overridden span is used).

The issue with `emit_unescape_error` doing arithmetic on arbitrary spans still remain.

estebank commented:
@petrochenkov addressed your comments and see that you fixed the underlying issue. While I was in this file I updated the diagnostics to bring the up to our current standard of quality.

petrochenkov commented on compiler/rustc_parse/src/lexer/unescape_error_reporting.rs:
```diff
@@ -102,21 +119,22 @@ pub(crate) fn emit_unescape_error(
+            } else {
+                diag.help(
+                    "for more information, visit \
+                     <https://static.rust-lang.org/doc/master/reference.html#literals>",
```
In which cases we link to the reference from error messages?
In theory, there's a reference page for every error, but in practice the diagnostics are already too noisy, so perhaps it's not a good idea to add such links just in case, without a strong specific reason.

estebank replied:
The case where this one triggers when using a "fake" escape, like `\y`. We could enumerate all the possible ones that are valid, but feared that would be too verbose. Every other case handled in this document provides enough information inline. I'm more worried about the link being stale. The reason I'm adding it is because these escapes are hard to search for.

petrochenkov commented on compiler/rustc_parse/src/lexer/unescape_error_reporting.rs:
```diff
@@ -29,9 +31,9 @@ pub(crate) fn emit_unescape_error(
+        let lo = span.lo() + BytePos(start);
         let hi = lo + BytePos(end - start);
-        span_with_quotes.with_lo(lo).with_hi(hi)
+        span.with_lo(lo).with_hi(hi)
```
This code is still doing arithmetic on arbitrary spans.
`let span = self.mk_sp(content_start, content_end)` will turn `span` into `override_span` in exactly the same way as `span_with_quotes`.

All the range arithmetic should be done on `content_start`/`content_end` which are guaranteed to be from lexer, and `mk_sp` should only be called in the end on its results (and it may drop those results and replace them with `override_span`).

estebank replied:
Moved this logic to the caller to operate on `content_*` directly.

petrochenkov commented on compiler/rustc_parse/src/lexer/mod.rs:
```diff
@@ -525,17 +523,26 @@ impl<'a> StringReader<'a> {
+        content_start: BytePos,
+        content_end: BytePos,
+        prefix: u32,
+        postfix: u32,
```
```suggestion
        prefix_len: u32,
        postfix_len: u32,
```

petrochenkov commented:
The PR name and description need an update.

bors commented:
:umbrella: The latest upstream changes (presumably #81625) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

petrochenkov commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit 3b5d018ebb6ee8aa6181ebf1cf9ba0cb407f622f has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 3b5d018ebb6ee8aa6181ebf1cf9ba0cb407f622f -->
<!-- homu: {"type":"Approved","sha":"3b5d018ebb6ee8aa6181ebf1cf9ba0cb407f622f","approver":"petrochenkov"} -->
